pkg/controllers/provisioningv2/rke2/common.go

```diff
@@ -403,3 +404,76 @@ func SortedKeys(m map[string]interface{}) []string {
 	sort.Strings(keys)
 	return keys
 }
+
+type Getter[T any] interface {
+	Get(namespace, name string) (*T, error)
+}
+
+var errNilObject = errors.New("cannot get capi cluster for nil object")
+
+// FindCAPIClusterFromLabel takes an obj T and will attempt to find the label denoting which capi cluster it belongs to.
+// If the object is nil, it cannot access to object or type metas, or the label is not present, it returns an error.
+func FindCAPIClusterFromLabel[T any](obj *T, getter Getter[capi.Cluster]) (*capi.Cluster, error) {
```

Comment:
@jakefhyde is the intention here to be able to retrieve the labels on "any" K8s-like object to look for the CAPI cluster label value?

We have the boilerplate for functionality similar to this (in a generic fashion) with the wrangler data package, see: https://github.com/rancher/wrangler/blob/master/pkg/data/data.go#L29 that operates on runtime.Object -- where are we sourcing the objects that are passed into these functions? I'm not sure that it makes sense to use generics here.

Reply:
Is this better?
```go
func FindCAPIClusterFromLabel(obj runtime.Object, getter Getter[capi.Cluster]) (*capi.Cluster, error) {
	data, err := data.Convert(obj)
	if err != nil {
		return nil, err
	}
	if data == nil {
		return nil, errNilObject
	}
	clusterName := data.String("metadata", "labels", capi.ClusterLabelName)
	if clusterName != "" {
		return getter.Get(data.String("metadata", "namespace"), clusterName)
	}
	return nil, fmt.Errorf("%s label not present on %s: %s/%s", capi.ClusterLabelName, obj.GetObjectKind().GroupVersionKind().Kind, data.String("metadata", "namespace"), data.String("metadata", "name"))
}
```
The only reason I was using generics here is because `if obj == nil` is always false (the interface is never nil, but the underlying pointer value is). 

Example usages can be found here: https://github.com/rancher/rancher/blob/8c101f5f63dc103e7ecc5476a08491994a9fc32d/pkg/controllers/provisioningv2/rke2/machineprovision/controller.go#L437
and here:
https://github.com/rancher/rancher/blob/8c101f5f63dc103e7ecc5476a08491994a9fc32d/pkg/controllers/provisioningv2/rke2/machineprovision/controller.go#L448